airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/write/object_storage/ObjectStorageStreamLoaderFactory.kt

```diff
@@ -79,28 +78,20 @@ class ObjectStorageStreamLoader<T : RemoteObject<*>, U : OutputStream>(
 ) : StreamLoader {
     private val log = KotlinLogging.logger {}
 
-    // Used for naming files. Distinct from part index, which is used to track uploads.
-    private val fileNumber = AtomicLong(0L)
     private val objectAccumulator = PartToObjectAccumulator(stream, client)
 
-    override suspend fun start() {
-        val state = destinationStateManager.getState(stream)
-        // This is the number used to populate {part_number} on the object path.
-        // We'll call it file number here to avoid confusion with the part index used for uploads.
-        val fileNumber = state.getNextPartNumber()
-        log.info { "Got next file number from destination state: $fileNumber" }
-        this.fileNumber.set(fileNumber)
-    }
-
     override suspend fun createBatchAccumulator(): BatchAccumulator {
+        val state = destinationStateManager.getState(stream)
         return RecordToPartAccumulator(
             pathFactory,
             bufferedWriterFactory,
             partSizeBytes = partSizeBytes,
             fileSizeBytes = fileSizeBytes,
             stream,
-            fileNumber
-        ) { name -> destinationStateManager.getState(stream).ensureUnique(name) }
+            state.getPartIdCounter(pathFactory.getFinalDirectory(stream)),
```

Comment:
Given the state has both the pathFactory and the stream, I wonder if `pathFactory.getFinalDirectory(stream)` should just happen internally in the state.